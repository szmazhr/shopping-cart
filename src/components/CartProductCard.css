.cart-product-card{
  display: grid;
  grid-template-columns: 300px 1fr 200px;
  gap: 1rem;
  max-width: 1440px;
  margin: auto;
  width: 100%;
  padding: 0 2rem;
  align-items: center;
  padding: 1rem;
  justify-items: center;
  box-shadow: 0 0 2px 0 rgba(0,0,0,0.5);
  transition: box-shadow 0.3s ease-in-out;
}
.cart-product-card:hover{
  box-shadow: 0 0 5px 0 rgba(0,0,0,0.5);
}
.cart-product-card .cart-product-card-img{
  display: flex;
  justify-content: center;
  height: 100%;
}
.cart-product-card .cart-product-card-img img{
  object-fit: cover;
  max-height: 200px;
  width: 300px;
}
.cart-product-card .cart-product-card-info{
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: 100%;
  height: 100%;
  gap: 15px;
  flex: 1;
}
.cart-product-card .cart-product-card-info a{
 text-decoration: none;
 color: rgb(27, 27, 27);
}
.cart-product-card .cart-product-card-info .cart-product-card-title{
  font-size: 2rem;
  font-weight: bold;
  flex: 1;
}
.cart-product-card .cart-product-card-info a:hover{
  text-decoration: underline;
 }
.cart-product-card .cart-product-card-info h3{
  margin: 0;
}

.cart-product-card .cart-product-card-quantity{
  display: flex;
  align-items: center;
  width: 180px;
}
.cart-product-card .cart-product-card-quantity label{
  flex: 1;
  font-weight: 700;
}
.cart-product-card .cart-product-card-quantity input{
  width: 3rem;
  text-align: center;
}
.cart-product-card .cart-product-card-btn{
  padding: .5rem;
  text-align: center;
  border: none;
  font-size: 1.2rem;
  color: rgb(27, 27, 27);
  cursor: pointer;
}
.cart-product-card .cart-product-card-btn:hover{
filter: brightness(0.9);
}

.cart-product-card button.cart-product-card-btn{
  background-color: rgb(226, 226, 226);
}
