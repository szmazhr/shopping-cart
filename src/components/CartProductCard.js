import { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import './CartProductCard.css';


function CartProductCard({ product, removeItem, setQuantity }) {
  
  const [count, setCount] = useState(product.quantity);

  const countIncrement = () => {
    setCount(count => count + 1);

  }

  const countDecrement = () => {
    if(count > 1) {
    setCount(count => count - 1);
    setQuantity(product.id, count);
    }
  }

  useEffect(() => {
    setQuantity(product.id, count);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [count] );



  const changeHandler = (e) => {
    const value = e.target.value;
    if(value > 0 && value.match(/^[0-9]*$/)) {
      setCount(e.target.value);
    }
  }
  return (
    <div className="cart-product-card">
    <div className="cart-product-card-img">
      <img src={product.image} loading="lazy" alt="product" />
    </div>
    <div className="cart-product-card-info">
      <Link to={`/product/${product.id}`}>
        <h3 className="cart-product-card-title">{product.title}</h3>
      </Link>
      <div className='cart-product-card-quantity'>
        <label htmlFor="product-quantity">Quantity: </label>
        <button type='button' disabled={count > 1 ? false : true } onClick={countDecrement} className="quantity-btn">-</button>
        <input type="text" value={count} id="product-quantity" onChange={changeHandler} />
        <button type='button' onClick={countIncrement} className="quantity-btn">+</button>
      </div>

      {/* <Link to='/cart' className="cart-product-card-btn">View Cart</Link> */}
    </div>
      <div className="action-btns">
        
      <p className="cart-product-card-price"><strong>Amount:</strong> ${product.price * product.quantity}</p>

      <button className="cart-product-card-btn" onClick={()=> removeItem(product.id)}>Remove</button>
      </div>
  </div>
  )
}

export default CartProductCard