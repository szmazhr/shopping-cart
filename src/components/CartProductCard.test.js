import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import CartProductCard from "./CartProductCard";

describe("CartProductCard", () => {

  const removeItem = jest.fn();
  const setQuantity = jest.fn();
  const product = {
                  id: 1,
                  title: "Product 1",
                  image: "https://via.placeholder.com/300",
                  price: 10,
                  quantity: 1
                }

  it("renders correctly", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <CartProductCard
          product={product}
          removeItem={() => {}}
          setQuantity={() => {}}
        />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders correctly with Quantity 5", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <CartProductCard
          product={{...product, quantity: 5}}
          removeItem={() => {}}
          setQuantity={() => {}}
        />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('calls removeItem when remove button is clicked', () => {
    render(
      <MemoryRouter>
        <CartProductCard
          product={product}
          removeItem={removeItem}
          setQuantity={() => {}}
        />
      </MemoryRouter>
    );

    userEvent.click(screen.getByText('Remove'));
    expect(removeItem).toHaveBeenCalledTimes(1);
    expect(removeItem).toHaveBeenCalledWith(product.id);
  });
  it('calls setQuantity when quantity input is increment', () => {
    render(
      <MemoryRouter>
        <CartProductCard
          product={product}
          removeItem={() => {}}
          setQuantity={setQuantity}
        />
      </MemoryRouter>
    );
    userEvent.click(screen.getByText('+'));
    expect(setQuantity).toHaveBeenCalledTimes(2); // one is for the initial value, one for the increment
    expect(setQuantity).toHaveBeenCalledWith(product.id, 2);
  });
  it('calls setQuantity when quantity input is decrement', () => {
    render(
      <MemoryRouter>
        <CartProductCard
          product={product}
          removeItem={() => {}}
          setQuantity={setQuantity}
        />
      </MemoryRouter>
    );

    userEvent.click(screen.getByText('-'));
    expect(setQuantity).toHaveBeenCalledTimes(1); // one is for the initial value, not for the decrement forbidden below 1
    expect(setQuantity).toHaveBeenCalledWith(product.id, 1);
  })
  it('calls setQuantity when quantity input is typed', () => {
    render(
      <MemoryRouter>
        <CartProductCard
          product={product}
          removeItem={() => {}}
          setQuantity={setQuantity}
        />
      </MemoryRouter>
    );
    userEvent.type(screen.getByLabelText(/^Quantity/i), '2');
    expect(setQuantity).toHaveBeenCalledTimes(2);
    expect(setQuantity).toHaveBeenCalledWith(product.id, 1);
    expect(setQuantity).toHaveBeenCalledWith(product.id, '12');
  });
});