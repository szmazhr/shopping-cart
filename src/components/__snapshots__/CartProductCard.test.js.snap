// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartProductCard renders correctly 1`] = `
<DocumentFragment>
  <div
    class="cart-product-card"
  >
    <div
      class="cart-product-card-img"
    >
      <img
        alt="product"
        loading="lazy"
        src="https://via.placeholder.com/300"
      />
    </div>
    <div
      class="cart-product-card-info"
    >
      <a
        href="/product/1"
      >
        <h3
          class="cart-product-card-title"
        >
          Product 1
        </h3>
      </a>
      <div
        class="cart-product-card-quantity"
      >
        <label
          for="product-quantity"
        >
          Quantity: 
        </label>
        <button
          class="quantity-btn"
          disabled=""
          type="button"
        >
          -
        </button>
        <input
          id="product-quantity"
          type="text"
          value="1"
        />
        <button
          class="quantity-btn"
          type="button"
        >
          +
        </button>
      </div>
    </div>
    <div
      class="action-btns"
    >
      <p
        class="cart-product-card-price"
      >
        <strong>
          Amount:
        </strong>
         $10
      </p>
      <button
        class="cart-product-card-btn"
      >
        Remove
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CartProductCard renders correctly with Quantity 5 1`] = `
<DocumentFragment>
  <div
    class="cart-product-card"
  >
    <div
      class="cart-product-card-img"
    >
      <img
        alt="product"
        loading="lazy"
        src="https://via.placeholder.com/300"
      />
    </div>
    <div
      class="cart-product-card-info"
    >
      <a
        href="/product/1"
      >
        <h3
          class="cart-product-card-title"
        >
          Product 1
        </h3>
      </a>
      <div
        class="cart-product-card-quantity"
      >
        <label
          for="product-quantity"
        >
          Quantity: 
        </label>
        <button
          class="quantity-btn"
          type="button"
        >
          -
        </button>
        <input
          id="product-quantity"
          type="text"
          value="5"
        />
        <button
          class="quantity-btn"
          type="button"
        >
          +
        </button>
      </div>
    </div>
    <div
      class="action-btns"
    >
      <p
        class="cart-product-card-price"
      >
        <strong>
          Amount:
        </strong>
         $50
      </p>
      <button
        class="cart-product-card-btn"
      >
        Remove
      </button>
    </div>
  </div>
</DocumentFragment>
`;
